i6_models/parts/losses/nce.py

```diff
@@ -0,0 +1,111 @@
+__all__ = [
+    "NoiseContrastiveEstimationLossV1",
+]
+
+import torch
+from torch import nn
+from torch.nn import functional as F
+from typing import Optional
+
+
+class NoiseContrastiveEstimationLossV1(nn.Module):
+    __constants__ = ["num_samples", "log_norm_term", "reduction"]
+    num_samples: int
+    log_norm_term: float
+
+    def __init__(
+        self,
+        num_samples: int,
+        *,
+        model: nn.Module,
+        noise_distribution_sampler: nn.Module,
+        log_norm_term: Optional[float] = None,
+        reduction: str = "none",
+        device: Optional[str] = None,
+    ) -> None:
+        """
+        Noise contrastive estimation loss implementation.
+        Used to estimate the softmax. Normally for very large softmax sizes, for example word-level LM.
+
+        :param num_samples: num of samples for the estimation, normally a value between 1000-4000.
+                            2000 is a good starting point.
+        :param model: model on which the NCE loss is to be applied. The model requires a member called `output`, which
+            is expected to be a linear layer with bias and weight members. The member `output` represents the output
+            layer of the model allowing access to the parameters during loss computation.
+        :param noise_distribution_sampler: for example `i6_model.samplers.LogUniformSampler`.
+        :param log_norm_term: normalisation term for true/sampled logits.
+        :param reduction: reduction method for binary cross entropy.
+        :param device: device where the loss will be placed.
+        """
+        super().__init__()
+
+        self.num_samples = num_samples
+        self.model = model  # only used to access weights of output layer for NCE computation
+        self.noise_distribution_sampler = noise_distribution_sampler
+        self.log_norm_term = log_norm_term
+        self.device = device
+
+        self._bce = nn.BCEWithLogitsLoss(reduction=reduction)
+
+    def forward(self, data: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        """
+
+        :param data: the tensor for the input data, where batch and time are flattened resulting in [B x T, F] shape.
```

Comment:
The comments here are more confusing than helpful. So it should be specified what data is (or even rename it), First I assumed it is logits, but then I guess this should be the input to the "output" layer of the provided model? This is very confusing and unclear. Then it should also be clear what is the input shape. When you say "where batch and time are flattened resulting in" this sounds like this is done internally, but what you mean it expects [B x T, F] as only valid input, correct?

Reply:
Are the new docs clearer?